Docs/Debugging/printf.md

```diff
@@ -1,106 +1,231 @@
+
+You can learn more on [Wikipedia](http://en.wikipedia.org/wiki/Printf_format_string).
+
+## Printf() from an interrupt context
+
+If you run the following code, you'll be in for a nasty surprise:
+
+```cpp
+#include "mbed.h"
+
+DigitalOut led(LED1);
+InterruptIn btn(SW2);
+
+void do_something() {
+  led = !led;
+  printf("Toggle LED!\r\n"); // CRASH! Blocking call in ISR...
+}
+
+int main() {
+  btn.fall(&do_something);
+
+  while (1) { }
+}
+```
+
+Your board crashes right when you press the button. This is because calls to stdio functions (like printf) are [guarded by mutexes](https://developer.mbed.org/handbook/CMSIS-RTOS) in the ARM C standard library, and mutexes [cannot be called from an ISR](https://www.keil.com/pack/doc/cmsis/RTOS/html/group__CMSIS__RTOS__MutexMgmt.html). 
+
+You can get around this by:
+
+* Signaling from the ISR to the main thread using a [Semaphore](https://docs.mbed.com/docs/mbed-os-api-reference/en/latest/APIs/tasks/rtos/#semaphore) or [Mailbox](https://docs.mbed.com/docs/mbed-os-api-reference/en/latest/APIs/tasks/rtos/#mail), and doing the `printf` call in the main thread.
+* Using an event dispatching library like [mbed Events](https://docs.mbed.com/docs/mbed-os-api-reference/en/latest/APIs/tasks/events/).
+
+Example code for both approaches are in [this blog post](https://developer.mbed.org/blog/entry/Simplify-your-code-with-mbed-events/).
+
+## Printf() macros
+
+You can use macro-replacement (performed by the pre-proccessor) to do some nifty tricks with ``printf()``.
 
 The general form for a simple macro definition is:
 
-	#define MACRO_NAME value 
+```
+#define MACRO_NAME value
+```
 
-This associates with the **MACRO_NAME** whatever **value** appears between the first space after the **MACRO_NAME** and the end of the line. The value constitutes the body of the macro.
+This associates the macro's name (`MACRO_NAME`) with the body of the macro, represented by whatever `value` appears between the first space after the `MACRO_NAME` and the end of the line.
 
-``printf()``s are very useful for debugging when looking for an explanation to a problem. Otherwise, it is nice to be able to disable many of them. We can use the ``#define`` directive to create parameterized macros that extend the basic ``printf()`` functionality. For example, macros can expand to printf()s when needed, but to empty statements under other conditions. 
+``printf()`` is very useful for debugging when looking for an explanation to a problem. Otherwise, it is nice to be able to disable many of them. We can use the ``#define`` directive to create parameterized macros that extend the basic ``printf()`` functionality. For example, macros can expand to printf()s when needed, but to empty statements under other conditions.
```

Comment:
Query: What does "them" refer to in "it is nice to be able to disable many of them"?